### ISSUE TITLE
Make the usage of msg_sender() a bit more ergonomic.

### ISSUE THREAD

**Issue by nfurfaro** [FuelLabs/sway#1309]
In Ethereum development, `msg.sender` is used everywhere for authentication. We simply call msg.sender and know that we’ll get an `Address` back.
We can’t  do that in Sway because:
- `Address` and `ContractId` are different types
- either type may be returned by `msg_sender()`
- an error may also be returned (if coins are spent from more than one address, then a unique sender can't be determined)

We can at least abstract most of the boilerplate needed to unwrap a nested Result away from the user for the most common cases. 

Open to input on how we do this. Probably the most helpful thing would be simple examples (even pseudocode) of how you would like to be able to use `msg_sender()` in practice, from which we can derive the best way forward.

## Ideas for helper functions
- similar to `is_err()` and `is_ok()`, we could have `is_address()` & `is_contract()` which return `bool` values.
- we can have an `unwrap_address()` and `unwrap_contract()` which either resolve to the thing or panic, or return Options. 
- We could even have a `unwrap_bytes()` which returns the inner `b256` value of either an `Address` or a `ContractId`

The problem with the Options is that it just defers the actual unwrapping and pushes it back into the users code, which is sort of what I want to avoid. That said, I think we should maybe do both so users can decide how to handle the absence of the value they want (ie: panic (revert) or something else). 

For many use cases, panicking when an expected address isn't there makes sense. We encode these sorts of checks everywhere when developing on Ethereum, ie: `Require(msg.sender != 0x234.., "Wrong Address!");`
so this would be the equivalent of saying `Assert(msg_sender().is_address);` before proceeding to use the Address value.

Note that enum matching (coming soon) will make this much cleaner, but those should mostly be implementation details most users don't need to care/know about.

cc @SilentCicero @Braqzen

**Comment by adlerjohn** [comment]
`unwrap_XXX` shouldn't return an `Option`, as that's congruent to how `unwrap` work. It should simply panic if trying to unwrap the wrong of the two variants.

I'm not sure it's necessarily good to over-egonomicize without actual user complaints, **especially when a proper solution is around the corner**. There's literally no end to how much more abstract you can make a flow, making more and more assumptions the more abstract it gets. Implementing actual features is more critical.

**Comment by nfurfaro** [comment]
Here's a rough draft of what I'm thinking. Keep in mind that the internals of all of these can be swapped out to make use of enum matching when it lands which should clean them up.

Not sure if there are any valid use cases for `unwrap_sender_bytes()`... 
Note that this is all untested, typed out in a few minutes for bikeshedding.

```
///////////////////////////////////////////
/// msg_sender helpers
///////////////////////////////////////////

/// Returns whether the result is a Address, letting the calling code decide how to handle the return value (true/false or error).
pub fn is_address(result: Result<Sender, AuthError>) -> bool {
    if result.is_ok() {
        if let Sender::Address(a) = e {
            true
        } else {
            false
        }
    } else {
        false
    }
}

/// Panics on Error or ContractId, otherwise returns an Address
/// Use when you expect an Address && a ContractId is not a valid value to continue executiion with.
pub fn unwrap_address(result: Result<Sender, AuthError>) -> Address {
    if let Sender::Address(a) = result.unwrap() {
        a
    } else {
        panic(0)
    }
}

/// Returns Option::Some(Address), or None.
/// Use this when only an Address is valid, but you don't neccessarily want to panic.
pub fn unwrap_address_or_none(result: Result<Sender, AuthError>) -> Option<Address> {
    if is_address(result) {
        if let Sender::Address(a) = result.unwrap() {
            Option::Some(a)
        } else {
            Option::None
        }
    } else {
        Option::None
    }
}

/// Returns whether the result is a Address, letting the calling code decide how to handle the return value (true/false or error).
pub fn is_contract_id(result: Result<Sender, AuthError>) -> bool {
    if result.is_ok() {
        if let Sender::ContractId(a) = e {
            true
        } else {
            false
        }
    } else {
        false
    }
}

/// Panics on Error or Address, otherwise returns an ContractId
/// Use when you expect a ContractId && an Address is not a valid value to continue executiion with.
pub fn unwrap_contract_id(result: Result<Sender, AuthError>) -> ContractId {
    if let Sender::ContractId(c) = result.unwrap() {
        c
    } else {
        panic(0)
    }
}



/// Returns Option::Some(ContractId), or None.
/// Use this when only an ContractId is valid, but you don't neccessarily want to panic.
pub fn unwrap_contract_id_or_none(result: Result<Sender, AuthError>) -> Option<ContractId> {
    if is_contract_id(result) {
        if let Sender::ContractId(c) = result.unwrap() {
            Option::Some(c)
        } else {
            Option::None
        }
    } else {
        Option::None
    }
}



/// Unwraps a Result and either returns the inner b256 value (from a ContractID or Address), or panics.
pub fn unwrap_sender_bytes(result: Result<Sender, AuthError>) -> b256 {
    if let Sender::Address(a) = result.unwrap() {
        a.value
    } else {
        if let Sender::ContractId(c) = result.unwrap() {
            c.value
        }
    }
}

```

**Comment by nfurfaro** [comment]
@adlerjohn just read your comment above. Yes we have a "proper" solution coming in the way of enum matches, but ideally most users won't even need to use them for handling the vast majority of the times they use msg_sender().

**Comment by adlerjohn** [comment]
Other than being more wasteful in gas and providing less flexibility and more confusion due to a larger API, in what way is `sender.unwrap_address()` less verbose than `sender.unwrap().address()`? The latter provides you with an intermediate `Sender` that you can then `match` over once that's available in the language, if you don't do the `.address(). The former does not.

I don't actually see any of the proposals as more ergonomic as simply `sender.unwrap()` followed by an if-else. The real way to make this ergonomic is `match`.

**Comment by adlerjohn** [comment]
> either type may be returned by msg_sender()

I think my issue with the proposed changes is that this statement is incorrect. `msg_sender()` does not return an `Address` or a `ContractId`. It returns a `Result`. Results are unwrapped if you're fine panicking on the `Err` variant.

**Comment by Braqzen** [comment]
I've read over the slack comments from a few days ago and the comments here.
My stance is that we should wait for matching so that the user can match against the sender types (address or contract) or the auth error.

It could be something like
```
let variable = match msg_sender() {
    // cases where address or contract are acceptable
    // auth error -> panic()
}
```

This pattern will likely be common so you may want to add helper functions where one returns an `Address` (`get_address()`) and the other helper returns `ContractID` (`get_contract()`).
I dislike proposing this but both of them would panic if they cannot return their expected type.

This would make a clear distinction for which function can be interacted with via a user or contract.
If it is user only and you call in with a contract then `get_address()` would panic.

This means that
1. The user can do custom matching when that is implemented
2. The user can use the helper functions `get_address()` and `get_contract()` instead of `msg_sender()`

I kind of like the idea of having an option to use something like `is_address()` and `is_contract()` and then use that in my own control flow too although this may mean that most people would not use this if the "get_XXX" functions are implemented. I would use the "is_XXX" calls to probably just panic if they are not the correct type ... so why even support them? Are there cases that enough users would want that counteract the "get_XXX" usage?

I also dislike the idea of maintaining additional functions that just unwrap code. I think that should be left to the user in conjunction with the functions above. As long as we make the functions safe then they can use `.unwrap()` and it should "just work"

P.S. can we refer to `Address` as address and `ContractID` as contract?
They seem to only be namespaced. When I think of getting the address or contract ... "id" I do not think of getting a "contract ID". I think of getting an address that is specifically typed as a contract.

**Comment by SilentCicero** [comment]
So while I'm all about making things more ergonomic.

TLDR: I think we need to maybe wait a little for other patterns to emerge as @adlerjohn says.

### My two cents:

I think one thing is that: Ethereum only has an "address" so developers **are not use to differentiating between a contract and address**. I found this a little confusing and hard to martial as you kind of have an "or" in the result.

**Either it "results" or "errors" if it results, it "could be an address" or it "could be a contract". This was particularly odd to manage as a return.** There are a few edges there. Ideally many design patterns don't care, so long as they can get an address like Ethereum.

I think trying to clean up the logical control pattern for the above might be more important than handy methods. In a sense, I'd just like to be able to get an Address, and *not care* about whether it's a contract or not. That was something I ran into before.

msg_sender -> address might be nice for that (where determining whether it's a contract or not is maybe too verbose).

FWIW put all those "handy" methods in a personal repo, and we can install them and see which methods we end up using in the future!

I'd do this with SwaySwaps, address_or_panic for example.

**Comment by Braqzen** [comment]
I'm absolutely in favour of having `msg_sender()` return "an address" and not care about the type however my proposition was to treat both the address and contractID as the same type and that was stated to not be a viable path atm (if ever).

**Comment by adlerjohn** [comment]
> I think trying to clean up the logical control pattern for the above might be more important than handy methods. In a sense, I'd just like to be able to get an Address, and not care about whether it's a contract or not.

That already exists. It's `Sender`! The solution would then be to make things like "transfer coins" accept a `Sender` rather than have two separate functions to transfer to a contract or transfer to an address.

**Comment by nfurfaro** [comment]
For me, the desire for an easier to use `msg_sender()` definitely stems from being used to always getting an `Address` on Ethereum, and now making the switch to handling either an `Address` or a `ContractId` or an `Error`.
Happy to park this for now, and when enum matching arrives see what common patterns emerge. 

For writing tests, I'll probably continue to use some helpers along these lines as we've already seen it can make a mess of the code to have multiple nested if lets every time you want to use msg_sender.

As for the statement "either type may be returned by msg_sender()", yes, this is technically not true and I recant it.

>I'd just like to be able to get an Address, and not care about whether it's a contract or not 

>you may want to add helper functions where one returns an Address (get_address()) and the other helper returns ContractID (get_contract()).
I dislike proposing this but both of them would panic if they cannot return their expected type.

That's exactly what the proposed `unwrap_address()` & `unwrap_contract_id()` do; return the expected type or panic.

>in what way is sender.unwrap_address() less verbose than sender.unwrap().address()? 

It's not? But I don't think I ever suggested it was...

This would be nice in many ways, and I guess the question is, is the "type safety" gained by having separate `Address` and  `ContractId` types worth the added friction it causes for things like msg_sender and transferring coins? 
I think it most likely is, but that we still have work to do to extend the usage of Enum types like Sender. 
I found myself wanting to find a better name for Sender that could be more generic; an enum which can be either an Address or ContractId, but could be used to name a Sender or a recipient of coins, etc...  while still being correct (ie: not just using Address for everything). I settled on "Entity" in the short term but I don't love it.

I do think we should consider making better use of the Sender enum (as mentioned above) so that users don't need to care about what type of sender/recipient they're dealing with.

**Comment by nfurfaro** [comment]
My thoughts on the naming so far:
`Sender` or `Caller` are both fine in that they describe a generic(one of 2 types) source or origin for the call. But the naming falls short if we want to generalize this concept of either type, ie: to pass it into a function that transfers coins to either an Address or ContractId. Now we're describing a destination / recipient which is either an Address or ContractId.

I think we need a single name for an enum identical to the current `Sender` but which semantically can include both the origin/source of an call, _and_ the endpoint/destination of a call. 

If both a `Sender` and a `Recipient` can be either an `Address` or a `ContractId`, what name to we give a thing which can replace both?

So far, `Entity` is the only name I've thought of which meets these criteria, but definitely open to ideas...

**Comment by adlerjohn** [comment]
I see your point, that makes sense. `Entity` is definitely a bit obtuse but kind of makes sense...let's see if we can think of something cuter.

**Comment by Braqzen** [comment]
```
enum System {
    sender: Entity,
    recipient: Entity
}

enum Entity {
    address: Address,
    contract: ContractID
}
```

**Comment by nfurfaro** [comment]
Operator is not bad... better than Entity maybe.
but, there's some baggage from the Ethereum world with regards to the meaning of Operator. 🤔

**Comment by SilentCicero** [comment]
@adlerjohn I think it would be nicer if we could clarify the sender construction a little bit.

I would prefer a data structure that has the single Address item I care about, but specifies whether the address is a Contract or Account.

The way the Sender struct is currently put together means, it's hard for me to determine "is it a contract or address", given that its only a tuple with two properties.

I think a better construction would be something where it tells me "this is a contract or address" then I can determine the address from it.

I found handling the Sender very clumsy. I had to go through all this logic to determine what I wanted which is an address.

If the struct or enum just told me, pick ContractId or pick Address that would be better IMO.

**Comment by SilentCicero** [comment]
On the topic of Entity, it's too abstract. Caller is likely more accurate and clear.

**Comment by SilentCicero** [comment]
> I'm absolutely in favour of having `msg_sender()` return "an address" and not care about the type however my proposition was to treat both the address and contractID as the same type and that was stated to not be a viable path atm (if ever).

The problem with this is then you lose the data of "is it a contract" or not. Which can be important. I think Johns idea to have a more complex type that is used which retains that data would be nice.

Transfer should accept a Sender like primitive. And making a Sender like primitive should be straight forward.

The only issue is, there are access control situations where "you don't know" if it's going to be a contract or not. And we should have type or way to state this as well.

**Comment by nfurfaro** [comment]
>I think a better construction would be something where it tells me "this is a contract or address" then I can determine the address from it.

I feel like the thing we're discussing (an enum with 2 variants, ie Sender) could achieve this?
you could do (pseudocode): 
```
let sender = msg_sender();
match sender {
    Address(a) => ... //do something with the address,
    ContractId(c) => ... //do something with the ContractId.
```
The type is encoded already, and as long as we use a type that could be either, you'll always have to perform at least some operation to learn which type it actually is.

**Comment by nfurfaro** [comment]
What about

```
pub enum Identity {
    Address: Address,
    ContractId: ContractId,
}
```

We can ensure that all commonly used functionality works with this type out of the box, so `msg_sender()` would return one inside the result, and a single transfer function could handle sending to either a contract or an output.
And users can always unwrap or otherwise extract the inner value manually if they need to. 
So this would basically be like `Sender` is now but used more widely and consistently, ie: when sending coins.

**Comment by adlerjohn** [comment]
> [Caller is likely more accurate and clear.](https://github.com/FuelLabs/sway/issues/1309#issuecomment-1104532597)

No. Please re-read @nfurfaro's comment: https://github.com/FuelLabs/sway/issues/1309#issuecomment-1104093227

> [I would prefer a data structure that has the single Address item I care about, but specifies whether the address is a Contract or Account.](https://github.com/FuelLabs/sway/issues/1309#issuecomment-1104533613)

That is literally exactly what `Sender` is. Please review Rust enums: https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html

**Comment by simonr0204** [comment]
I'm a little late to this conversation, but my 2 cents regarding specifically:

> Ethereum only has an "address" so developers are not use to differentiating between a contract and address

This is true, although in Solidity on those occasions where you _do_ need to find out what you're dealing with, you have to open an assembly block and query `EXTCODESIZE` which is... not great. It seems well within the spirit of Sway to require the developer to explicitly handle both cases, (or otherwise explicitly decline to do so).

**Comment by SwayStar123** [comment]
> What about
> 
> ```
> pub enum Identity {
>     Address: Address,
>     ContractId: ContractId,
> }
> ```
> 
> We can ensure that all commonly used functionality works with this type out of the box, so `msg_sender()` would return one inside the result, and a single transfer function could handle sending to either a contract or an output. And users can always unwrap or otherwise extract the inner value manually if they need to. So this would basically be like `Sender` is now but used more widely and consistently, ie: when sending coins.

How about identity traits which can just be implemented for both address and contractid?

**Comment by nfurfaro** [comment]
>How about identity traits which can just be implemented for both address and contracted?

@SwayStar123 Could you give an example of what you're thinking this would look like?

**Comment by SwayStar123** [comment]
> > How about identity traits which can just be implemented for both address and contracted?
> 
> @SwayStar123 Could you give an example of what you're thinking this would look like?

```
trait Identity {
    fn receive(&self) -> u64;
}

impl Identity for Address {
    fn receive(&self) -> u64 {
    // code to send assets to this address
    }
}

impl Identity for Sender {
    fn receive(&self) -> u64 {
        match self { //do the address one if its an address, do the contractid one if its a contractid }
   }
}
```

something like that
